devicetypes/shinasys/sihas-multipurpose-sensor.src/sihas-multipurpose-sensor.groovy

```diff
@@ -0,0 +1,297 @@
+        translatable   : true
+    ]
+}
+
+private Map parseIasMessage(String description) {
+    ZoneStatus zs = zigbee.parseZoneStatus(description)
+
+    translateZoneStatus(zs)
+}
+
+private Map translateZoneStatus(ZoneStatus zs) {
+    // Some sensor models that use this DTH use alarm1 and some use alarm2 to signify motion
+    return (zs.isAlarm1Set() || zs.isAlarm2Set()) ? getMotionResult('active') : getMotionResult('inactive')
+}
+
+private Map getBatteryResult(rawValue) {
+    def linkText = getLinkText(device)
+
+    def result = [:]
+
+    def volts = rawValue / 10
+
+    if (!(rawValue == 0 || rawValue == 255)) {
+        result.name = 'battery'
+        result.translatable = true
+        def minVolts =  2.3
+        def maxVolts =  3.2
+        // Get the current battery percentage as a multiplier 0 - 1
+        def curValVolts = Integer.parseInt(device.currentState("battery")?.value ?: "100") / 100.0
+        // Find the corresponding voltage from our range
+        curValVolts = curValVolts * (maxVolts - minVolts) + minVolts
+        // Round to the nearest 10th of a volt
+        curValVolts = Math.round(10 * curValVolts) / 10.0
+        // Only update the battery reading if we don't have a last reading,
+        // OR we have received the same reading twice in a row
+        // OR we don't currently have a battery reading
+        // OR the value we just received is at least 2 steps off from the last reported value
+        if (state?.lastVolts == null || state?.lastVolts == volts || device.currentState("battery")?.value == null || Math.abs(curValVolts - volts) > 0.1) {
+            def pct = (volts - minVolts) / (maxVolts - minVolts)
+            def roundedPct = Math.round(pct * 100)
+            if (roundedPct <= 0)
+                roundedPct = 1
+            result.value = Math.min(100, roundedPct)
+        } else {
+            // Don't update as we want to smooth the battery values, but do report the last battery state for record keeping purposes
+            result.value = device.currentState("battery").value
+        }
+        result.descriptionText = "${device.displayName} battery was ${result.value}%"
+        state.lastVolts = volts
+    }
+
+    return result
+}
+
+private Map getMotionResult(value) {
+    String descriptionText = value == 'active' ? "${device.displayName} detected motion" : "${device.displayName} motion has stopped"
+    return [
+        name           : 'motion',
+        value          : value,
+        descriptionText: descriptionText,
+        translatable   : true
+    ]
+}
+
+/**
+ * PING is used by Device-Watch in attempt to reach the Device
+ * */
+def ping() {
+    zigbee.readAttribute(zigbee.POWER_CONFIGURATION_CLUSTER, POWER_CONFIGURATION_BATTERY_VOLTAGE_ATTRIBUTE)
+}
+
+def refresh() {
+    def refreshCmds = []
+
+    refreshCmds += zigbee.readAttribute(zigbee.POWER_CONFIGURATION_CLUSTER, POWER_CONFIGURATION_BATTERY_VOLTAGE_ATTRIBUTE)
+    if (isUSM300() || isTSM300())
+    {
```

Comment:
Please  move this opening bracket to the line above, so it would be: if (...) {
Applies to the whole file.